Source/Core/VideoCommon/TextureInfo.cpp

```diff
@@ -0,0 +1,325 @@
+    }
+    break;
+  case 256 * 2:
+  {
+    for (size_t i = 0; i < m_texture_size; i++)
+    {
+      const u32 texture_byte = m_data[i];
+
+      min = std::min(min, texture_byte);
+      max = std::max(max, texture_byte);
+    }
+    break;
+  }
+  case 16384 * 2:
+    for (size_t i = 0; i < m_texture_size; i += sizeof(u16))
+    {
+      const u32 texture_halfword = Common::swap16(m_data[i]) & 0x3fff;
+
+      min = std::min(min, texture_halfword);
+      max = std::max(max, texture_halfword);
+    }
+    break;
+  }
+  if (tlut_size > 0)
+  {
+    tlut_size = 2 * (max + 1 - min);
+    tlut += 2 * min;
+  }
+
+  const u64 tex_hash = XXH64(m_data, m_texture_size, 0);
+  const u64 tlut_hash = tlut_size ? XXH64(tlut, tlut_size, 0) : 0;
+
+  m_base_name = fmt::format("{}{}x{}{}_{:016x}", format_prefix, m_raw_width, m_raw_height,
+                            m_mip_levels.empty() ? "" : "_m", tex_hash);
+  m_tlut_name = tlut_size ? fmt::format("_{:016x}", tlut_hash) : "";
+  m_format_name = fmt::format("_{}", static_cast<int>(m_texture_format));
+
+  m_name_cached = true;
+}
+
+std::optional<std::string> TextureInfo::GetFullName() const
+{
+  if (!m_name_cached)
+    return std::nullopt;
+
+  return m_base_name + m_tlut_name + m_format_name;
+}
+
+std::string_view TextureInfo::GetBaseName() const
+{
+  return m_base_name;
+}
+
+std::string_view TextureInfo::GetTlutName() const
+{
+  return m_tlut_name;
+}
+
+std::string_view TextureInfo::GetFormatName() const
+{
+  return m_format_name;
+}
+
+const u8* TextureInfo::GetData() const
+{
+  return m_data;
+}
+
+const u8* TextureInfo::GetTlutAddress() const
+{
+  return m_tlut_address;
+}
+
+u32 TextureInfo::GetRawAddress() const
+{
+  return m_raw_address;
+}
+
+bool TextureInfo::IsFromTmem() const
+{
+  return m_from_tmem;
+}
+
+const u8* TextureInfo::GetTmemOddAddress() const
+{
+  return m_tmem_odd;
+}
+
+TextureFormat TextureInfo::GetTextureFormat() const
+{
+  return m_texture_format;
+}
+
+TLUTFormat TextureInfo::GetTlutFormat() const
+{
+  return m_tlut_format;
+}
+
+std::optional<u32> TextureInfo::GetPaletteSize() const
+{
+  return m_palette_size;
+}
+
+u32 TextureInfo::GetTextureSize() const
+{
+  return m_texture_size;
+}
+
+u32 TextureInfo::GetBlockWidth() const
+{
+  return m_block_width;
+}
+
+u32 TextureInfo::GetBlockHeight() const
+{
+  return m_block_height;
+}
+
+u32 TextureInfo::GetExpandedWidth() const
+{
+  return m_expanded_width;
+}
+
+u32 TextureInfo::GetExpandedHeight() const
+{
+  return m_expanded_height;
+}
+
+u32 TextureInfo::GetRawWidth() const
+{
+  return m_raw_width;
+}
+
+u32 TextureInfo::GetRawHeight() const
+{
+  return m_raw_height;
+}
+
+bool TextureInfo::HasMipMaps() const
+{
+  return !m_mip_levels.empty();
+}
+
+u32 TextureInfo::GetLevelCount() const
+{
+  return static_cast<u32>(m_mip_levels.size()) + 1;
+}
+
+const TextureInfo::MipLevel* TextureInfo::GetMipMapLevel(u32 level) const
+{
+  if (m_mip_levels.size() > level)
+    return &m_mip_levels.at(level);
```

Comment:
```suggestion
    return &m_mip_levels[level];
```

Given the bounds is already checked before this access.

Reply:
Yeah, the reason this is done is this is a `const` function and `[]` is not const.

Reply:
(I do the check because Dolphin doesn't support exceptions, so I think this would be a hard crash??)

Reply:
There are both const and non-const versions of `[]`: https://en.cppreference.com/w/cpp/container/array/operator_at (one returning a `T&` and the other returning `const T&`).

Reply:
But not for `std::vector` until C++20:  https://en.cppreference.com/w/cpp/container/vector/operator_at

Reply:
Ah.  I was thinking of map.  Great, I'll switch back to `[]`.  Thanks all